File: app/dtos/challenge.py
from datetime import date, datetime

from pydantic import BaseModel

from app.dtos.base import BaseSerializerModel
from app.models.challenge import (
    ChallengeCategory,
    ChallengeTrack,
    CheckinEmotion,
    UserChallengeStatus,
)


class ChallengeResponse(BaseSerializerModel):
    id: int
    name: str
    category: ChallengeCategory
    description: str
    duration_days: int
    track: ChallengeTrack
    stage: int  # 난이도 단계 (1=입문 2=초보 3=중급 4=숙련)


class ChallengeListResponse(BaseSerializerModel):
    total: int
    items: list[ChallengeResponse]


class JoinChallengeRequest(BaseModel):
    challenge_id: int
    started_at: date


class UserChallengeResponse(BaseSerializerModel):
    id: int
    challenge_id: int
    started_at: date
    status: UserChallengeStatus
    streak_count: int
    total_checkins: int
    last_checkin_date: date | None
    created_at: datetime


class UserChallengeListResponse(BaseSerializerModel):
    total: int
    items: list[UserChallengeResponse]


class HeatmapDay(BaseSerializerModel):
    date: date  # YYYY-MM-DD
    count: int  # 그 날 체크인 횟수


class HeatmapResponse(BaseSerializerModel):
    weeks: int  # 표시 주 수 (26)
    today: date
    days: list[HeatmapDay]  # weeks*7개 (오래된 날짜부터)
    max_count: int  # 색상 단계 결정용


class CategoryProgress(BaseSerializerModel):
    category: ChallengeCategory
    percent: int  # 0~100
    active_count: int  # 해당 카테고리 활성 챌린지 수
    total_checkins: int
    total_duration: int


class CategoryProgressResponse(BaseSerializerModel):
    items: list[CategoryProgress]  # 5종 (HYDRATION/EXERCISE/DIET/SLEEP/STRESS)


class CheckinRequest(BaseModel):
    emotion: CheckinEmotion | None = None  # 선택, 그 날의 감정


class EmotionDay(BaseSerializerModel):
    date: date
    emotion: CheckinEmotion | None  # 그 날 기록된 감정 (없으면 None)


class WeeklyEmotionResponse(BaseSerializerModel):
    days: list[EmotionDay]  # 최근 7일


class CheckinAwardResponse(BaseSerializerModel):
    base: int
    lucky: bool
    lucky_extra: int
    streak_bonus: int
    streak_milestone: int
    full_participation: bool
    full_participation_bonus: int
    total: int


class EggUpdateResponse(BaseSerializerModel):
    progress_checkins: int
    current_stage: int  # 0=알, 1=부화, 2/3/4=진화 단계
    goal_70_just_alerted: bool
    goal_90_just_alerted: bool
    stage_bonus: int
    stage_milestone: int  # 도달한 임계 (10/40/100/200)
    hatched: bool  # 1단계 부화 (종 추첨 시점)
    evolved_to: int | None  # 진화한 단계 번호 (2/3/4), 부화일 땐 None
    is_legendary: bool | None
    species: str | None
    character_name: str | None
    new_egg_no: int | None


class CheckInResponse(BaseSerializerModel):
    id: int
    streak_count: int
    total_checkins: int
    last_checkin_date: date
    status: UserChallengeStatus
    message: str
    award: CheckinAwardResponse | None = None
    egg: EggUpdateResponse | None = None


class CancelCheckinResponse(BaseSerializerModel):
    id: int
    streak_count: int
    total_checkins: int
    last_checkin_date: date | None
    status: UserChallengeStatus
    points_revoked: int
    message: str


class AbandonChallengeResponse(BaseSerializerModel):
    id: int
    status: UserChallengeStatus
    points_revoked: int
    message: str


# ─── 챌린지 재설계: 트랙/스테이지/필수체크 응답 DTO ───────────────────────────


class TrackCategoryInfo(BaseSerializerModel):
    """트랙에 속한 카테고리 정보 (UI 탭 목록용)"""

    category: ChallengeCategory
    label: str  # 한글 라벨


class MyTrackResponse(BaseSerializerModel):
    """내 트랙 정보 조회 응답"""

    track: ChallengeTrack
    track_label: str
    stage: int
    stage_label: str
    auto_assigned: bool
    categories: list[TrackCategoryInfo]  # 트랙의 카테고리 목록 (UI 탭)


class UpdateMyTrackRequest(BaseModel):
    """배지 단계(stage) 변경 요청.

    트랙은 PDF 명세상 자동배정되며 사용자가 변경할 수 없으므로 stage만 받는다.
    """

    stage: int  # 1~4


class DailyChecklistItemResponse(BaseSerializerModel):
    """일일 필수체크 항목 개별 응답"""

    item_key: str
    text: str
    checked: bool


class DailyChecklistResponse(BaseSerializerModel):
    """일일 필수체크 전체 응답"""

    date: date  # YYYY-MM-DD
    track: ChallengeTrack
    items: list[DailyChecklistItemResponse]


class ChecklistToggleResponse(BaseSerializerModel):
    """필수체크 항목 토글 응답 — 포인트·알 적립 결과 포함."""

    item_key: str
    text: str
    checked: bool
    points_awarded: int  # 이번 토글 순변동 (+5 / +35 / -5 / -35 / 0)
    all_completed: bool  # 토글 후 트랙 필수항목 전체완료 여부
    full_bonus_awarded: int  # 이번에 새로 지급된 전체완료 보너스 (0 또는 30)
    egg: EggUpdateResponse | None = None  # 전체완료로 알이 진행됐을 때만


# ─── 월별 달성 달력 DTO ───────────────────────────────────────────────────────


class CalendarDay(BaseSerializerModel):
    date: date
    required: bool
    selected_count: int  # 그날 체크인한 선택 챌린지 카테고리 종 수
    level: str  # none | basic | silver | gold


class MonthlyCalendarResponse(BaseSerializerModel):
    year_month: str  # YYYY-MM
    days: list[CalendarDay]  # 해당 월 1일~말일
    achieved_days: int  # level != none 일수
    gold_days: int  # level == gold 일수
    max_streak: int  # level != none 연속 최장 (월 내)

File: frontend/ckd-care-app/src/api/challenge.ts
import { api } from "./client";

// ── 신버전 타입: 5트랙·9카테고리·stage ──────────────────────────────────────
export type ChallengeCategory =
  | "HYDRATION"
  | "EXERCISE"
  | "DIET"
  | "SLEEP"
  | "STRESS"
  | "EDUCATION"
  | "RECORD"
  | "MONITORING"
  | "EMOTION";

export type ChallengeTrack =
  | "DIALYSIS"
  | "CKD"
  | "INTENSIVE"
  | "DAILY"
  | "WELLNESS";

// CKD 진단자 트랙 — 이 트랙이면 챌린지 화면을 진단자 전용(서브탭)으로 분기
export const CKD_TRACKS: ChallengeTrack[] = ["CKD", "DIALYSIS"];

export type UserChallengeStatus = "ACTIVE" | "COMPLETED" | "ABANDONED";

export interface Challenge {
  id: number;
  name: string;
  category: ChallengeCategory;
  description: string;
  duration_days: number;
  track: ChallengeTrack;
  stage: number; // 1~4
}

// ── 기존 유지 타입 ────────────────────────────────────────────────────────────
export interface ChallengeListResponse {
  total: number;
  items: Challenge[];
}

export interface UserChallenge {
  id: number;
  challenge_id: number;
  started_at: string;
  status: UserChallengeStatus;
  streak_count: number;
  total_checkins: number;
  last_checkin_date: string | null;
  created_at: string;
}

export interface UserChallengeListResponse {
  total: number;
  items: UserChallenge[];
}

export interface CheckinAward {
  base: number;
  lucky: boolean;
  lucky_extra: number;
  streak_bonus: number;
  streak_milestone: number;
  full_participation: boolean;
  full_participation_bonus: number;
  total: number;
}

export interface EggUpdate {
  progress_checkins: number;
  current_stage: number;  // 0=알, 1=부화, 2/3/4=진화 단계
  goal_70_just_alerted: boolean;
  goal_90_just_alerted: boolean;
  stage_bonus: number;
  stage_milestone: number;  // 도달한 임계 (10/40/100/200)
  hatched: boolean;  // 1단계 부화 (종 추첨)
  evolved_to: number | null;  // 진화한 단계 (2/3/4)
  is_legendary: boolean | null;
  species: string | null;
  character_name: string | null;
  new_egg_no: number | null;
}

export interface CheckInResponse {
  id: number;
  streak_count: number;
  total_checkins: number;
  last_checkin_date: string;
  status: UserChallengeStatus;
  message: string;
  award: CheckinAward | null;
  egg: EggUpdate | null;
}

export interface HeatmapDay {
  date: string;
  count: number;
}

export interface HeatmapResponse {
  weeks: number;
  today: string;
  days: HeatmapDay[];
  max_count: number;
}

export interface CategoryProgress {
  category: ChallengeCategory;
  percent: number;
  active_count: number;
  total_checkins: number;
  total_duration: number;
}

export interface CategoryProgressResponse {
  items: CategoryProgress[];
}

export interface CancelCheckinResponse {
  message: string;
  points_revoked: number;
}

// ── 신규 타입: my-track·daily-checklist ──────────────────────────────────────
export interface TrackCategoryInfo {
  category: ChallengeCategory;
  label: string;
}

export interface MyTrack {
  track: ChallengeTrack;
  track_label: string;
  stage: number;
  stage_label: string;
  auto_assigned: boolean;
  categories: TrackCategoryInfo[];
}

export interface DailyChecklistItem {
  item_key: string;
  text: string;
  checked: boolean;
}

export interface DailyChecklistResponse {
  date: string;
  track: ChallengeTrack;
  items: DailyChecklistItem[];
}

export interface DailyChecklistToggleResult {
  item_key: string;
  text: string;
  checked: boolean;
  points_awarded: number;   // 이번 토글 순변동 (+5/+35/-5/-35/0)
  all_completed: boolean;
  full_bonus_awarded: number; // 0 또는 30
  egg: EggUpdate | null;
}

// ── 월별 달력 타입 ────────────────────────────────────────────────────────────
export type CalendarLevel = "none" | "basic" | "silver" | "gold";

export interface CalendarDay {
  date: string; // YYYY-MM-DD
  required: boolean;
  selected_count: number;
  level: CalendarLevel;
}

export interface MonthlyCalendarResponse {
  year_month: string;
  days: CalendarDay[];
  achieved_days: number;
  gold_days: number;
  max_streak: number;
}

// ── challengeApi ─────────────────────────────────────────────────────────────
export const challengeApi = {
  // ── 신버전 (트랙·스테이지·필수체크) ─────────────────────────────────────
  myTrack: () => api.get<MyTrack>("/challenges/my-track"),
  // 트랙은 자동배정되어 변경 불가 — 배지 단계(stage)만 변경한다.
  updateMyTrack: (stage: number) =>
    api.put<MyTrack>("/challenges/my-track", { stage }),
  dailyChecklist: () => api.get<DailyChecklistResponse>("/challenges/daily-checklist"),
  toggleChecklist: (itemKey: string) =>
    api.post<DailyChecklistToggleResult>(`/challenges/daily-checklist/${itemKey}`, {}),
  listByTrackStage: (track: ChallengeTrack, stage: number) =>
    api.get<ChallengeListResponse>(`/challenges?track=${track}&stage=${stage}`),
  // ── 기존 유지 (참여·체크인·게이미피케이션) ──────────────────────────────
  myList: (limit = 100, offset = 0) =>
    api.get<UserChallengeListResponse>(`/user-challenges?limit=${limit}&offset=${offset}`),
  join: (challenge_id: number, started_at: string) =>
    api.post<UserChallenge>("/user-challenges", { challenge_id, started_at }),
  checkin: (userChallengeId: number) =>
    api.post<CheckInResponse>(`/user-challenges/${userChallengeId}/checkin`, {}),
  cancelCheckin: (userChallengeId: number) =>
    api.delete<CancelCheckinResponse>(`/user-challenges/${userChallengeId}/checkin`),
  abandon: (userChallengeId: number) =>
    api.delete<{ message: string }>(`/user-challenges/${userChallengeId}`),
  heatmap: (weeks = 26) => api.get<HeatmapResponse>(`/challenges/heatmap?weeks=${weeks}`),
  categoryProgress: () => api.get<CategoryProgressResponse>("/challenges/category-progress"),
  weeklyEmotion: () => api.get<WeeklyEmotionResponse>("/challenges/weekly-emotion"),
  calendar: (yearMonth?: string) =>
    api.get<MonthlyCalendarResponse>(`/challenges/calendar${yearMonth ? `?year_month=${yearMonth}` : ""}`),
};

// ── 감정 이모지 (기존 유지) ────────────────────────────────────────────────
export type CheckinEmotion =
  | "VERY_HAPPY"
  | "HAPPY"
  | "NEUTRAL"
  | "ANXIOUS"
  | "SAD"
  | "ANGRY"
  | "TIRED";

export const EMOTION_EMOJI: Record<CheckinEmotion, string> = {
  VERY_HAPPY: "😄",
  HAPPY: "🙂",
  NEUTRAL: "😐",
  ANXIOUS: "😟",
  SAD: "😢",
  ANGRY: "😠",
  TIRED: "😴",
};

export interface EmotionDay {
  date: string;
  emotion: CheckinEmotion | null;
}

export interface WeeklyEmotionResponse {
  days: EmotionDay[];
}
